src/Lean/Server/FileWorker.lean

```diff
@@ -164,178 +216,133 @@ structure WorkerState where
       : IO (WorkerContext × WorkerState) := do
     let clientHasWidgets := initParams.initializationOptions?.bind (·.hasWidgets?) |>.getD false
-    let (headerStx, headerTask) ← compileHeader meta o opts (hasWidgets := clientHasWidgets)
-    let cancelTk ← CancelToken.new
+    let mut mainModuleName := Name.anonymous
+    try
+      if let some path := System.Uri.fileUriToPath? meta.uri then
+        mainModuleName ← moduleNameOfFileName path none
+    catch _ => pure ()
+    let maxDocVersionRef ← IO.mkRef 0
+    let chanOut ← mkLspOutputChannel maxDocVersionRef
+    let srcSearchPathPromise ← IO.Promise.new
+
+    let processor := Language.Lean.process (setupImports meta chanOut srcSearchPathPromise)
+    let processor ← Language.mkIncrementalProcessor processor { opts, mainModuleName }
+    let initSnap ← processor meta.mkInputContext
     let ctx := {
-      hIn  := i
-      hOut := o
+      chanOut
       hLog := e
-      headerTask
       initParams
+      processor
       clientHasWidgets
+      maxDocVersionRef
+      opts
+    }
+    let doc : EditableDocumentCore := {
+      meta, initSnap
+      diagnosticsRef := (← IO.mkRef ∅)
     }
-    let cmdSnaps ← EIO.mapTask (t := headerTask) (match · with
-      | Except.ok (s, _) => unfoldCmdSnaps meta #[s] cancelTk ctx (startAfterMs := 0)
-      | Except.error e   => throw (e : ElabTaskError))
-    let doc : EditableDocument := { meta, cmdSnaps := AsyncList.delayed cmdSnaps, cancelTk }
+    let reporter ← reportSnapshots ctx doc
     return (ctx, {
-      doc                := doc
-      initHeaderStx      := headerStx
-      currHeaderStx      := headerStx
-      importCachingTask? := none
+      doc := { doc with reporter }
+      srcSearchPathTask  := srcSearchPathPromise.result
       pendingRequests    := RBMap.empty
       rpcSessions        := RBMap.empty
+      importCachingTask? := none
     })
+  where
+    /-- Creates an LSP message output channel along with a reader that sends out read messages on
+        the output FS stream after discarding outdated notifications. This is the only component of
+        the worker with access to the output stream, so we can synchronize messages from parallel
+        elaboration tasks here. -/
+    mkLspOutputChannel maxDocVersion : IO (IO.Channel JsonRpc.Message) := do
+      let chanOut ← IO.Channel.new
+      let _ ← chanOut.forAsync (prio := .dedicated) fun msg => do
+        -- discard outdated notifications; note that in contrast to responses, notifications can
+        -- always be silently discarded
+        let version? : Option Nat := do
+          let doc ← match msg with
+            | .notification "textDocument/publishDiagnostics" (some <| .obj params) => some params
+            | .notification "$/lean/fileProgress" (some <| .obj params) =>
+              params.find compare "textDocument" |>.bind (·.getObj?.toOption)
+            | _ => none
+          doc.find compare "version" |>.bind (·.getNat?.toOption)
```

Comment:
I would prefer doing some regular parsing to the respective param type here.

Reply:
I don't really understand the need for the Json/Structured distinction I have to say, but after overcoming that confusion it doesn't look too bad with parsing

Reply:
https://www.jsonrpc.org/specification#conventions